src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs

```diff
@@ -32,48 +32,53 @@ protected override void EndProcessing()
         {
             string filePath = null;
             string tempPath = Path.GetTempPath();
-            int attempts = 0;
-            bool temporaryFilePathFound = false;
-            while(attempts++ < 10 && !temporaryFilePathFound)
-            {
-                string fileName = Path.GetRandomFileName();
-                fileName = Path.ChangeExtension(fileName, _extension);
-                filePath = Path.Combine(tempPath, fileName);
-                if(!File.Exists(filePath))
-                {
-                    temporaryFilePathFound = true;
-                }
-            }
-            if(!temporaryFilePathFound)
-            {
-                WriteError(
-                    new ErrorRecord(
-                            new IOException("Could not find an available temporary file name in {tempPath}."),
-                            "NewTemporaryFileResourceUnavailable",
-                            ErrorCategory.ResourceUnavailable,
-                            tempPath));
-                    return;
-            }
 
-            if (ShouldProcess(filePath))
+            if (ShouldProcess(tempPath))
             {
-                try
+                int attempts = 0;
+                bool creationOfFileSuccessful = false;
+
+                // In case the random temporary file already exists, retry 
+                while (attempts++ < 10 && !creationOfFileSuccessful)
                 {
-                    using (new FileStream(filePath, FileMode.CreateNew)) { }
-                    FileInfo file = new FileInfo(filePath);
-                    WriteVerbose("Created temporary file {file.FullName}.");
-                    WriteObject(file);
+                    try
+                    {
+                        string fileName = Path.GetRandomFileName();
+                        fileName = Path.ChangeExtension(fileName, _extension);
+                        filePath = Path.Combine(tempPath, fileName);
+                        using (new FileStream(filePath, FileMode.CreateNew)) { }
+                        creationOfFileSuccessful = true;
+                        WriteVerbose("Created temporary file {filePath}.");
+                    }
+                    catch (IOException) // file already exists -> retry
+                    {
+                        attempts++;
+                    }
+                    catch (Exception exception) // fatal error, which could be e.g. insufficient permissions, etc.
+                    {
+                        WriteError(
+                        new ErrorRecord(
+                                exception,
+                                "NewTemporaryFileWriteError",
+                                ErrorCategory.WriteError,
+                                tempPath));
+                        return;
+                    }
                 }
-                catch (Exception e)
+
+                if (!creationOfFileSuccessful)
                 {
                     WriteError(
```

Comment:
Thanks for the quick turnaround. 

The following, which predates your PR, may not be worth doing, because it is technically a breaking change, but it's worth pointing out: 

Based on the official cmdlet error-reporting guidelines, this should be a (statement)-_terminating_ error: "Cmdlets that can accept only 0-1 input objects and generate only 0-1 output objects should treat errors as terminating errors and generate terminating exceptions." - https://msdn.microsoft.com/en-us/library/ms714412%28v=vs.85%29.aspx  

cc @SteveL-MSFT